userspace: Add --no-relax build option

ceolin opened the pull request:
In some architectures the linker performs global optimization relaxing address modes and changing instructions in the output object file. This is a problem when userspace is enabled since it assumes that addresses won't changes after certain build stage. In no supported architectures this option is ignored.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
kconfig: linker: Add --no-relax build option

In some architectures the linker performs global optimization relaxing
address modes and changing intructions in the output object file. This
is a problem when userspace is enabled since it assumes that addresses
won't change after certain build stage. In no supported architectures
this option is ignored.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

tejlmand reviewed (changes requested):
Adding `--no-relax` can reduce  performance, as the linker is no longer able to substiture long / in-effective jump calls to shorter / more effective instructions.

So we should definitely mention this decrease in performance here:
https://github.com/zephyrproject-rtos/zephyr/blob/02d5290e070d13be39722596a1df99c260bae972/arch/Kconfig#L277

and probable also mention this somewhere in the docs:
https://docs.zephyrproject.org/latest/kernel/usermode/index.html

Please try to find a suitable location for this piece of information in general docs.

ceolin commented:
> Adding `--no-relax` can reduce performance, as the linker is no longer able to substiture long / in-effective jump calls to shorter / more effective instructions.
> 
> So we should definitely mention this decrease in performance here:
> 
> https://github.com/zephyrproject-rtos/zephyr/blob/02d5290e070d13be39722596a1df99c260bae972/arch/Kconfig#L277
> 
> and probable also mention this somewhere in the docs: [docs.zephyrproject.org/latest/kernel/usermode/index.html](https://docs.zephyrproject.org/latest/kernel/usermode/index.html)
> 
> Please try to find a suitable location for this piece of information in general docs.

That is a good point, but the performance penalty in this case is eclipsed by all other userspace required changes. I think a a comment in the code (together with the explanation why this is need) is enough. I mean, it is needed, we can afford the linker changing address so we can consider it as part of userspace "package".

ceolin commented:
@tejlmand can you review it again ?

tejlmand reviewed (changes requested):
> but the performance penalty in this case is eclipsed by all other userspace required changes.

will it ? I'm not into the finer details on the userspace implementation, but my understanding is that naturally the multithreading itself is the costly part, the book keeping required for the userspace addition is not as costly. (I could of course be wrong on this)

>  I mean, it is needed, we can afford the linker changing address so we can consider it as part of userspace "package".

Of course we can afford it, we must afford.
No question on that.

My comment is more related to the fact that enabling userspace results in extra degradation of performance, and this PR just revealed that such info is missing.

Of course any feature requires processing, but when we specifically requests the linker to not do an optimization, then I still think we should be clear on that.
We cannot expect everyone to look into build files.

Thinking a bit more on this, and the fact that this is a linker optimization flag, then maybe a Kconfig for enabling (disabling) `--no-relax` is the best path forward.
That would be clean, and allow features (not only userspace) to enable the linker flag, when there is a need.
The userspace Kconfig feature can then do a `select` on the corresponding symbol.

Pros:
- Feature / flag not limited to userspace
- Having link optimization configurable similar to compiler optimization
- No need to clutter userspace with description related to link flags.

ceolin commented:
> > but the performance penalty in this case is eclipsed by all other userspace required changes.
> 
> will it ? I'm not into the finer details on the userspace implementation, but my understanding is that naturally the multithreading itself is the costly part, the book keeping required for the userspace addition is not as costly. (I could of course be wrong on this)
> 
> > I mean, it is needed, we can afford the linker changing address so we can consider it as part of userspace "package".
> 
> Of course we can afford it, we must afford. No question on that.
> 
> My comment is more related to the fact that enabling userspace results in extra degradation of performance, and this PR just revealed that such info is missing.
> 
> Of course any feature requires processing, but when we specifically requests the linker to not do an optimization, then I still think we should be clear on that. We cannot expect everyone to look into build files.
> 
> Thinking a bit more on this, and the fact that this is a linker optimization flag, then maybe a Kconfig for enabling (disabling) `--no-relax` is the best path forward. That would be clean, and allow features (not only userspace) to enable the linker flag, when there is a need. The userspace Kconfig feature can then do a `select` on the corresponding symbol.
> 
> Pros:
> 
> * Feature / flag not limited to userspace
> * Having link optimization configurable similar to compiler optimization
> * No need to clutter userspace with description related to link flags.

Fair enough. I've added a Kconfig option for it. The default is using --relax.

tejlmand commented on Kconfig.zephyr:
```diff
@@ -292,6 +292,17 @@ config LINKER_LAST_SECTION_ID_PATTERN
+	bool "Use --relax Linker option"
+	default y
+	help
+	  This option performs global optimizations that become possible when the linker resolves
+	  addressing in the program, such as relaxing address modes and synthesizing new
+	  instructions in the output object file. On platforms where this is not supported,
+	  `--relax' is accepted, but ignored.
+	  Disabling it can reduce performance, as the linker is no longer able to substiture long /
+	  in-effective jump calls to shorter / more effective instructions.
```
Very nice, good help text  :+1:

tejlmand commented on Kconfig.zephyr:
```diff
@@ -292,6 +292,17 @@ config LINKER_LAST_SECTION_ID_PATTERN
 	  The size of the pattern must not exceed 4 bytes.
 
+config LINKER_USE_RELAX
+	bool "Use --relax Linker option"
```
Please remove `use` as per guide lines:
https://docs.zephyrproject.org/latest/build/kconfig/tips.html#prompt-strings
> For a Kconfig symbol that enables a driver/subsystem FOO, consider having just “Foo” as the prompt, instead of “Enable Foo support” or the like.

On top of that, the `--relax` is a bit ld centric, maybe a title like:
```suggestion
	bool "Linker optimization of call addressing"
```

tejlmand commented on Kconfig.zephyr:
```diff
@@ -292,6 +292,17 @@ config LINKER_LAST_SECTION_ID_PATTERN
+	help
+	  This option performs global optimizations that become possible when the linker resolves
+	  addressing in the program, such as relaxing address modes and synthesizing new
+	  instructions in the output object file. On platforms where this is not supported,
+	  `--relax' is accepted, but ignored.
```
Minor nit, as Zephyr is working towards better 3rd party toolchain support then placing ld/lld refs here is nice.
For example:
```suggestion
	  instructions in the output object file.
	  For ld and lld, this enables `--relax`. On platforms where this is not supported,
	  `--relax' is accepted, but ignored.
```

tejlmand commented on Kconfig.zephyr:
```diff
@@ -292,6 +292,17 @@ config LINKER_LAST_SECTION_ID_PATTERN
+	  `--relax' is accepted, but ignored.
+	  Disabling it can reduce performance, as the linker is no longer able to substiture long /
+	  in-effective jump calls to shorter / more effective instructions.
+
```
We need a corresponding `LINKER_USE_NO_RELAX` setting.

This allows other settings to either enforce `relax` (removing the option of disabling it), as well as allowing setting to force disabling of the setting.

```suggestion

config LINKER_USE_NO_RELAX
	bool
	help
	  Hidden symbol to allow features to force the use of no relax.

```

tejlmand commented on Kconfig.zephyr:
```diff
@@ -292,6 +292,17 @@ config LINKER_LAST_SECTION_ID_PATTERN
 
+config LINKER_USE_RELAX
+	bool "Use --relax Linker option"
+	default y
```
with the corresponding `NO_RELAX`, then a depends must be added here:
```suggestion
	depends on !LINKER_USE_NO_RELAX
	default y
```

tejlmand reviewed (changes requested):
Very nice.

We only need a `LINKER_USE_NO_RELAX` setting and then to update here:
https://github.com/zephyrproject-rtos/zephyr/blob/50274718e22a3757c79384e2b8a1f6c57d288a30/arch/Kconfig#L276
with a `select LINKER_USE_NO_RELAX` line to ensure this setting gets disabled when `USERSPACE` is enabled.

That will allow features to:
- Force the use of relaxation
- Force the use of no relaxation

while allowing users to enable or disable this feature in all other  cases.

tejlmand reviewed (changes requested):
If I understand the initial problem correctly then we want to enfurce the use of no relax when `CONFIG_USERSPACE=y`.

If that is the case, then you missed to do this comment:
> We only need a LINKER_USE_NO_RELAX setting and then to update here:
>
> https://github.com/zephyrproject-rtos/zephyr/blob/50274718e22a3757c79384e2b8a1f6c57d288a30/arch/Kconfig#L276
>
> with a `select LINKER_USE_NO_RELAX` line to ensure this setting gets disabled when `USERSPACE` is enabled.

Ref: https://github.com/zephyrproject-rtos/zephyr/pull/53237#pullrequestreview-1243567355

Commit:
userspace: Do not use --relax flag

In platforms where the linker is capable of doing global optimizations,
like relaxing address mode and synthesize new instructions, Zephyr has to
disable it when enabling USERSPACE since the build expects that address
don't change after the first stage build.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

ceolin commented:
> If I understand the initial problem correctly then we want to enfurce the use of no relax when `CONFIG_USERSPACE=y`.
> 
> If that is the case, then you missed to do this comment:
> 
> > We only need a LINKER_USE_NO_RELAX setting and then to update here:
> > https://github.com/zephyrproject-rtos/zephyr/blob/50274718e22a3757c79384e2b8a1f6c57d288a30/arch/Kconfig#L276
> > 
> > with a `select LINKER_USE_NO_RELAX` line to ensure this setting gets disabled when `USERSPACE` is enabled.
> 
> Ref: [#53237 (review)](https://github.com/zephyrproject-rtos/zephyr/pull/53237#pullrequestreview-1243567355)



> If I understand the initial problem correctly then we want to enfurce the use of no relax when `CONFIG_USERSPACE=y`.
> 
> If that is the case, then you missed to do this comment:
> 
> > We only need a LINKER_USE_NO_RELAX setting and then to update here:
> > https://github.com/zephyrproject-rtos/zephyr/blob/50274718e22a3757c79384e2b8a1f6c57d288a30/arch/Kconfig#L276
> > 
> > with a `select LINKER_USE_NO_RELAX` line to ensure this setting gets disabled when `USERSPACE` is enabled.
> 
> Ref: [#53237 (review)](https://github.com/zephyrproject-rtos/zephyr/pull/53237#pullrequestreview-1243567355)

:/ Oh man, forgot to push this. Thanks for checking it out !

tejlmand reviewed (approved):
lgtm, thanks for the effort :+1:

ceolin commented:
> lgtm, thanks for the effort +1

thank you !
